### PR TITLE
fix: exclude JVM-singleton empty collections from raw $ref path tracking

### LINKED ISSUE #7795: [BUG] fastjson1-compatible 序列化 Map<Integer, Bean> 中的共享 Collection 时生成缺少 Map key 的 $ref
### 问题描述

从 Fastjson 1.2.83 升级到 `fastjson1-compatible 2.0.63` 后，当根对象为 `Map<Integer, Bean>`，且多个 Bean 持有同一个 Collection 实例时，`JSON.toJSONString()` 会为后续 Collection 生成错误的 `$ref` 路径。

实际生成的引用：

```json
{
  "$ref": "$.ids"
}
```

该路径缺少根 Map 的 key。第一次出现的集合位于 `1001 -> ids`，如果需要输出引用，路径应该类似：

```json
{
  "$ref": "$[1001].ids"
}
```

使用 fastjson1-compatible 2.0.63 反序列化错误 `$ref` 时不会抛出异常，而是将后续 Bean 的 `ids` 字段静默设置为 `null`。

这里不是循环引用，而是多个 Bean 共享同一个 Collection 实例。

### 环境信息

- OS信息：macOS
- JDK信息：OpenJDK 17.0.16
- 版本信息：`com.alibaba:fastjson:2.0.63`（fastjson1-compatible）

### 重现步骤

运行以下代码：

```java
import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.TypeReference;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

public class FastjsonSharedCollectionRefTest {

    public static void main(String[] args) {
        Set<Integer> sharedEmptySet = Collections.emptySet();

        Map<Integer, Model> sourceModels = new LinkedHashMap<>();
        sourceModels.put(1001, new Model(sharedEmptySet));
        sourceModels.put(1002, new Model(sharedEmptySet));

        System.out.println(
                "Shared before serialization: "
                        + (sourceModels.get(1001).getIds()
                        == sourceModels.get(1002).getIds())
        );

        String json = JSON.toJSONString(sourceModels);
        System.out.println("Serialized JSON: " + json);

        Map<Integer, Model> parsedModels = JSON.parseObject(
                json,
                new TypeReference<Map<Integer, Model>>() {
                }
        );

        System.out.println("1001.ids: " + parsedModels.get(1001).getIds());
        System.out.println("1002.ids: " + parsedModels.get(1002).getIds());
    }

    public static class Model {

        private Set<Integer> ids;

        public Model() {
        }

        public Model(Set<Integer> ids) {
            this.ids = ids;
        }

        public Set<Integer> getIds() {
            return ids;
        }

        public void setIds(Set<Integer> ids) {
            this.ids = ids;
        }
    }
}
```

序列化结果中，第二个 `ids` 类似：

```json
{
  "$ref": "$.ids"
}
```

反序列化结果：

```text
Shared before serialization: true
1001.ids: []
1002.ids: null
```

### 期待的正确结果

兼容 Fastjson 1.2.83 的行为，两个集合均输出实际内容：

```json
{
  "1001": {
    "ids": []
  },
  "1002": {
    "ids": []
  }
}
```

或者，如果 fastjson1-compatible 需要使用引用检测，则 `$ref` 应包含第一次出现该集合的完整路径：

```json
{
  "$ref": "$[1001].ids"
}
```

反序列化后至少应满足：

```text
1001.ids: []
1002.ids: []
```

无法解析 `$ref` 时也不应静默返回 `null`，建议抛出明确的反序列化异常。

### 相关日志输出

没有异常日志。问题表现为反序列化成功，但第二个共享 Collection 字段被静默设置为 `null`。

错误引用：

```json
{
  "$ref": "$.ids"
}
```

该路径从根 `Map<Integer, Model>` 无法定位到第一次出现的集合，因为缺少 `1001` 这一层。

### 附加信息

使用以下方式关闭引用检测后可以正常工作：

```java
import com.alibaba.fastjson.serializer.SerializerFeature;

String json = JSON.toJSONString(
        sourceModels,
        SerializerFeature.DisableCircularReferenceDetect
);
```

此时两个 `ids` 均输出为 `[]`，反序列化后也不会变成 `null`。

相似问题：

- https://github.com/alibaba/fastjson2/issues/3516
- https://github.com/alibaba/fastjson2/issues/7793
- https://github.com/alibaba/fastjson2/issues/956

希望确认：

1. fastjson1-compatible 为共享 Collection 生成 `$ref` 是否属于预期兼容行为。
2. `$ref` 丢失外层 Map key 是否属于 writer bug。
3. 无法解析 `$ref` 时静默将字段设置为 `null` 是否可以改为抛出异常。

### PR DESCRIPTION
**Pull request by 1919chichi**
## Summary

Fixes #7795.

`Collections.emptySet()` / `emptyList()` / `emptyMap()` are JVM-wide singletons. `ObjectWriterProvider.isNotReferenceDetect()` is meant to exclude these (and other immutable value types) from `$ref` reference-detection, since tracking a JVM singleton by identity is meaningless. The guarded `JSONWriter.setPath()` / `popPath()` overloads already check this via `isRefDetect(Object)`.

However, ASM-generated bean field writers (`ObjectWriterCreatorASM`) call the **raw** `setPath0(FieldWriter, Object)` overload directly for non-`Object`-declared fields, bypassing that exclusion entirely on the push side. Meanwhile `popPath0(Object)` only special-cased `Collections.EMPTY_LIST`/`EMPTY_SET` by identity on the pop side (not `EMPTY_MAP`, and not any of the other excluded types reachable through this path). This push/pop asymmetry left `JSONWriter.path` one level too deep after writing a shared empty collection field, corrupting every path computed afterward.

Repro (works against plain `com.alibaba.fastjson2.JSON`, not just the fastjson1-compatible module):

```java
Set<Integer> sharedEmptySet = Collections.emptySet();
Map<Integer, Model> map = new LinkedHashMap<>();
map.put(1001, new Model(sharedEmptySet));
map.put(1002, new Model(sharedEmptySet));

JSON.toJSONString(map, JSONWriter.Feature.ReferenceDetection);
// before: {1001:{"ids":[]},1002:{"ids":{"$ref":"$.ids"}}}  <- bogus ref, path missing "1001"
// after:  {1001:{"ids":[]},1002:{"ids":[]}}
```

Deserializing the bogus `$ref` silently resolved to `null` instead of throwing or resolving to the real empty collection.

## Fix

Centralize the exclusion check in the raw choke-point methods on `JSONWriter` instead of patching every individual ASM codegen call site (which is scattered across ~5 methods in `ObjectWriterCreatorASM.java` and is how this asymmetry was introduced in the first place):

- `setPath0(FieldWriter, Object)`: early-return when `ObjectWriterProvider.isNotReferenceDetect(object.getClass())`.
- `setPath0(int, Object)`: same guard, for symmetry with the sibling raw overload (its one current caller, `ObjectWriterImplList.writeJSONB`, already pre-filters via a guarded check, so this is defense-in-depth on the public raw API rather than a currently-reachable fix).
- `popPath0(Object)`: replaced the `EMPTY_LIST`/`EMPTY_SET`-only identity check with the same class-based `isNotReferenceDetect` check, so push and pop are symmetric again (this also correctly extends the exclusion to `EMPTY_MAP`).

## Test plan

- [x] Added `core/src/test/java/com/alibaba/fastjson2/issues_7000/Issue7795.java`:
  - `testSharedEmptySetAcrossMapEntries` / `testSharedEmptyMapAcrossMapEntries`: reproduce the reported scenario (`Map<Integer, Bean>` with a shared empty `Set`/`Map` field); verified these fail without the `setPath0(FieldWriter, Object)` fix and pass with it.
  - `testSetPath0IntOverloadExcludesSingletons`: direct contract test for the `setPath0(int, Object)` overload.
- [x] `mvn -pl core -am validate` (checkstyle) clean.
- [x] `mvn -pl core -am test`: full core suite green (7977 tests, 0 failures/errors), both incrementally and after a clean rebuild.
- [x] Independently reviewed by two separate agents (one Claude-based, one Codex-based); both confirmed the root-cause diagnosis and fix are correct with no NPE risk or regression, and both independently caught a test-quality gap in an earlier draft (a `List`-based test that didn't actually exercise the fixed code path), which was reworked before this PR.

<details>
<summary>中文说明</summary>

## 概述

修复 #7795。

`Collections.emptySet()` / `emptyList()` / `emptyMap()` 是 JVM 全局单例，`ObjectWriterProvider.isNotReferenceDetect()` 的设计意图就是把它们（以及其他不可变值类型）从 `$ref` 引用检测中排除，因为按对象身份去追踪一个全局单例是没有意义的。有保护的 `JSONWriter.setPath()`/`popPath()` 已经通过 `isRefDetect(Object)` 做了这个检查。

但 ASM 生成的 Bean 字段写入代码（`ObjectWriterCreatorASM`）对于非 `Object` 声明类型的字段，会直接调用**裸的** `setPath0(FieldWriter, Object)`，在压栈侧完全绕过了这层排除；而 `popPath0(Object)` 在出栈侧只是按对象身份特判了 `Collections.EMPTY_LIST`/`EMPTY_SET` 两个单例（不包括 `EMPTY_MAP`，也不包括这条路径上其他本该排除的类型）。这种压栈会记、弹栈不弹的不对称，导致写完一个共享的空集合字段后，`JSONWriter.path` 多陷了一层且不再恢复，污染了后续所有路径计算。

## 修复方案

没有去逐个修补 `ObjectWriterCreatorASM.java` 里分散在约 5 处的 ASM 字节码生成调用点（这种分散正是这次不对称产生的原因），而是把排除判断集中收口到 `JSONWriter` 上的裸方法本身：

- `setPath0(FieldWriter, Object)`：命中 `ObjectWriterProvider.isNotReferenceDetect(object.getClass())` 时提前返回。
- `setPath0(int, Object)`：同样加上这层判断，保持和另一个裸重载的契约一致（它目前唯一的调用方 `ObjectWriterImplList.writeJSONB` 在调用前已经做了有保护的判断，所以这里属于对公开裸 API 的防御性加固，不对应当前可触发的路径）。
- `popPath0(Object)`：把原来只认 `EMPTY_LIST`/`EMPTY_SET` 两个单例的身份判断，换成同样的按 class 排除判断，让压栈和弹栈重新对称（顺带正确覆盖了 `EMPTY_MAP`）。

## 测试

新增 `core/src/test/java/com/alibaba/fastjson2/issues_7000/Issue7795.java`，并做了完整的正反验证：去掉修复代码时对应测试会失败、加回后通过；`mvn validate`（checkstyle）干净；`core` 模块全量测试 7977 个用例全部通过。修复方案本身也经过两个独立 agent（一个基于 Claude、一个基于 Codex）分别审查，结论一致：修复正确、无 NPE 风险、无回归；两边还各自独立发现了草稿阶段一个测试没有真正覆盖到修复点的问题，已在提交前修正。

</details>

## Review round 2 (4609730e6)

Addressing the review feedback on the raw path methods:

- **Null tolerance (Critical).** Centralizing the exclusion introduced an implicit non-null precondition — the previous implementations compared identity (`popPath0`) or only stored the value in an `IdentityHashMap` (`setPath0`), so `null` was harmless, whereas `object.getClass()` is not. This is reachable at runtime, not only via the public raw API: `ObjectWriterCreatorASM` guards the **push** side with an explicit `ifnull` before `setPath0` but emits the trailing `popPath0(fieldValue)` **unguarded**, so an ordinary bean with a `null` Collection field written as JSONB with `ReferenceDetection` + `WriteNulls` threw `NullPointerException` in `popPath0`.

  `null` now takes exactly the same path as an excluded class — **never pushed, never popped** — rather than restoring the old behavior verbatim. Restoring it would keep `popPath0(null)` popping a level `setPath0` never pushed, which is a pre-existing over-pop on this same path:

  ```java
  class Bean { List<String> aItems; Model b; Model c; }  // aItems == null, b == c
  root = {"k": bean}

  // before: JSONB -> {"k":{"aItems":null,"b":{...},"c":{"$ref":"$.b"}}}    <- missing ".k"
  //         text  -> {"k":{"aItems":null,"b":{...},"c":{"$ref":"$.k.b"}}}
  // after:  both  -> {"k":{"aItems":null,"b":{...},"c":{"$ref":"$.k.b"}}}
  ```

  Same push/pop asymmetry this PR is about, so it is fixed here rather than left behind.

- **Javadoc (Suggestion).** All three raw methods now state that they bypass only the `Feature.ReferenceDetection` feature check, still enforce the `ObjectWriterProvider.isNotReferenceDetect` class exclusion, ignore `null`, and apply identical exclusions on both sides.

- **Tests.** `Issue7795.testRawPathMethodsAcceptNull` (initializes `JSONWriter.path` first, otherwise the `path == null` guard masks the problem, then calls both `setPath0` overloads and `popPath0(null)`) and `Issue7795.testNullCollectionFieldDoesNotCorruptPath` (end-to-end JSONB/text, pinning the `$.k.b` ref above). Removing either `object == null` guard makes both fail with the NPE — verified. `mvn -pl core -am validate` clean; `mvn -pl core test` → 7979 tests, 0 failures, 0 errors.

<details>
<summary>中文说明（Review round 2）</summary>

- **null 兼容性（Critical）**：把排除判断收口后，隐含引入了「入参必须非空」的前提——旧实现要么做身份比较（`popPath0`），要么只是把值放进 `IdentityHashMap`（`setPath0`），`null` 都是安全的，而 `object.getClass()` 不是。这不只影响公开裸 API，运行时就能触发：`ObjectWriterCreatorASM` 在**压栈侧**用 `ifnull` 显式跳过 `setPath0`，但结尾的 `popPath0(fieldValue)` **没有保护**，因此一个带 `null` 集合字段的普通 Bean，在 `ReferenceDetection` + `WriteNulls` 下写 JSONB 会在 `popPath0` 抛 NPE。

  现在 `null` 走和被排除 class 完全相同的路径——**不压栈、也不弹栈**——而不是原样恢复旧行为。原样恢复会保留 `popPath0(null)` 弹掉一个 `setPath0` 从未压入的层级，这本身就是这条路径上已存在的多弹一层的问题（示例见上：JSONB 输出 `$.b`，文本 writer 对同一对象图输出 `$.k.b`）。这与本 PR 要修的是同一个压栈/弹栈不对称，故一并修复。

- **Javadoc（Suggestion）**：三个裸方法的 Javadoc 现在明确写明只绕过 `Feature.ReferenceDetection` 这层 feature 检查，仍执行 `ObjectWriterProvider.isNotReferenceDetect` 的 class 排除并忽略 `null`，且两侧排除条件完全一致。

- **测试**：新增 `testRawPathMethodsAcceptNull`（先初始化 `path`，否则 `path == null` 会掩盖问题）与 `testNullCollectionFieldDoesNotCorruptPath`（端到端，钉住上面的 `$.k.b`）。去掉任一 `object == null` 保护，两个测试都会以 NPE 失败（已验证）。`mvn -pl core -am validate` 干净；`mvn -pl core test` 7979 个用例全通过。

</details>

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01YFzJPj75rwUU5YetDZLSyd

### CODE REVIEW

File: core/src/main/java/com/alibaba/fastjson2/JSONWriter.java [hunk 1]
```diff
@@ -299,6 +299,10 @@ public final String setPath(FieldWriter fieldWriter, Object object) {
      * @return the previous path as a string, or null if no previous path exists
      */
     public final String setPath0(FieldWriter fieldWriter, Object object) {
+        if (ObjectWriterProvider.isNotReferenceDetect(object.getClass())) {
```
Comment (wenshao) [inline core/src/main/java/com/alibaba/fastjson2/JSONWriter.java]: **[Critical]** These raw path methods now dereference nullable values before applying the class exclusion. After a caller initializes `JSONWriter.path`, `setPath0(FieldWriter, null)`, `setPath0(int, null)`, or `popPath0(null)` reaches `object.getClass()` and throws, while the previous implementations accepted null. A probe against this PR failed at this line, and guarding the classifier with `object != null` made it pass. Please preserve the previous null behavior in all three changed methods.

```text
PR: JSONWriterNullProbeTest.rawPathMethodsAcceptNull -> NullPointerException
 at JSONWriter.setPath0(JSONWriter.java:302)
With null guard: Tests run: 1, Failures: 0, Errors: 0; BUILD SUCCESS
```

Add a regression test that calls both `setPath0` overloads and `popPath0(null)`; removing the null guard must make that test fail.

<details><summary>中文说明</summary>

这些裸 path 方法现在会在应用 class 排除判断之前解引用可能为 null 的值。调用方初始化 `JSONWriter.path` 后，`setPath0(FieldWriter, null)`、`setPath0(int, null)` 或 `popPath0(null)` 都会执行 `object.getClass()` 并抛出异常，而旧实现可以接受 null。针对当前 PR 的探针在此行失败，增加 `object != null` 保护后通过。请在三个改动的方法中都保留原有的 null 行为。并增加覆盖两个 `setPath0` 重载和 `popPath0(null)` 的回归测试；移除 null 保护时该测试必须失败。

</details>

_— gpt-5.6-sol via Qwen Code /review (v0.22.0)_

Comment (wenshao) [inline core/src/main/java/com/alibaba/fastjson2/JSONWriter.java]: **[Suggestion]** The public Javadocs for both `setPath0` overloads and `popPath0` still say they operate “without reference detection,” but these methods now skip every class rejected by `ObjectWriterProvider.isNotReferenceDetect`. That wording can lead callers to rely on a bypass that no longer exists. Please clarify that the raw methods bypass the feature-level check while still enforcing the unsupported-class exclusion.

<details><summary>中文说明</summary>

两个 `setPath0` 重载和 `popPath0` 的公开 Javadoc 仍然写着“without reference detection”，但这些方法现在会跳过 `ObjectWriterProvider.isNotReferenceDetect` 排除的所有 class。这个表述可能让调用方误以为仍能完全绕过引用检测条件。建议明确说明：裸方法只绕过 feature 层检查，但仍会执行不支持引用检测类型的排除。

</details>

_— gpt-5.6-sol via Qwen Code /review (v0.22.0)_

**Review by wenshao — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/alibaba/fastjson2/JSONWriter.java (hunk 1 above)
Comment (1919chichi) [reply core/src/main/java/com/alibaba/fastjson2/JSONWriter.java]: Fixed in 4609730e6.

Confirmed and reproduced — and it is worse than a raw-API-only concern: this NPE is reachable through ordinary serialization. `ObjectWriterCreatorASM` guards the **push** side with an explicit `ifnull` before `setPath0`, but emits the trailing `popPath0(fieldValue)` **unguarded** (`ObjectWriterCreatorASM.java:1137-1166`). So a plain bean with a `null` Collection field, written as JSONB with `ReferenceDetection` + `WriteNulls`, throws:

```text
java.lang.NullPointerException: Cannot invoke "Object.getClass()" because "object" is null
  at com.alibaba.fastjson2.JSONWriter.popPath0(JSONWriter.java:431)
  at com.alibaba.fastjson2.writer.OWG_1_2_Bean.writeJSONB(Unknown Source)
```

One deliberate deviation from the literal request: instead of restoring the old null behavior as-is, `null` now takes exactly the same path as an excluded class — **never pushed, never popped**. Restoring the old behavior would keep `popPath0(null)` popping a level that `setPath0` never pushed, which is a pre-existing over-pop bug on this same path:

```java
class Bean { List<String> aItems; Model b; Model c; }  // aItems == null, b == c
root = {"k": bean}

// before: JSONB -> {"k":{"aItems":null,"b":{...},"c":{"$ref":"$.b"}}}    <- missing ".k"
//         text  -> {"k":{"aItems":null,"b":{...},"c":{"$ref":"$.k.b"}}}
// after:  both  -> {"k":{"aItems":null,"b":{...},"c":{"$ref":"$.k.b"}}}
```

This is the same push/pop asymmetry the PR is about, so it is fixed here rather than left behind. Happy to split it out if you would rather keep this PR strictly scoped.

Regression tests added to `Issue7795`:
- `testRawPathMethodsAcceptNull` — initializes `JSONWriter.path` first (otherwise the `path == null` guard masks the problem), then calls both `setPath0` overloads and `popPath0(null)`.
- `testNullCollectionFieldDoesNotCorruptPath` — end-to-end JSONB/text, also pinning the `$.k.b` ref above.

Removing either `object == null` guard makes both tests fail with the NPE (verified). `mvn -pl core -am validate` clean; `mvn -pl core test` → 7979 tests, 0 failures, 0 errors.

<details><summary>中文说明</summary>

已在 4609730e6 修复。

问题确认，且比「仅影响裸 API」更严重：这个 NPE 通过普通序列化就能触发。`ObjectWriterCreatorASM` 在**压栈侧**用 `ifnull` 显式跳过了 `setPath0`，但结尾的 `popPath0(fieldValue)` 是**没有保护的**（`ObjectWriterCreatorASM.java:1137-1166`）。所以一个带 `null` 集合字段的普通 Bean，在 `ReferenceDetection` + `WriteNulls` 下写 JSONB 就会抛 NPE（栈见上）。

有一处刻意偏离原始建议：没有原样恢复旧的 null 行为，而是让 `null` 走和被排除 class 完全相同的路径——**不压栈、也不弹栈**。因为原样恢复会保留 `popPath0(null)` 弹掉一个 `setPath0` 从未压入的层级，这本身就是这条路径上已存在的多弹一层的 bug：JSONB 输出 `$.b`，而文本 writer 对同一对象图输出 `$.k.b`（示例见上）。这和本 PR 要修的是同一个压栈/弹栈不对称，所以一并修掉了。如果希望本 PR 严格控制范围，我可以拆分出去。

已在 `Issue7795` 补充两个回归测试：`testRawPathMethodsAcceptNull`（先初始化 `path`，否则 `path == null` 会掩盖问题，再分别调用两个 `setPath0` 重载和 `popPath0(null)`）和 `testNullCollectionFieldDoesNotCorruptPath`（端到端，同时钉住上面的 `$.k.b`）。去掉任一 `object == null` 保护，两个测试都会以 NPE 失败（已验证）。`mvn -pl core -am validate` 干净；`mvn -pl core test` 7979 个用例全通过。

</details>

Comment (1919chichi) [reply core/src/main/java/com/alibaba/fastjson2/JSONWriter.java]: Fixed in 4609730e6. All three Javadocs now state that the raw methods bypass only the `Feature.ReferenceDetection` feature check, while still enforcing the `ObjectWriterProvider.isNotReferenceDetect` class exclusion and ignoring `null` — and that this exclusion is identical on both sides, so a value that is not pushed by `setPath0` is not popped by `popPath0`.

<details><summary>中文说明</summary>

已在 4609730e6 修复。三个方法的 Javadoc 现在都明确写明：裸方法只绕过 `Feature.ReferenceDetection` 这层 feature 检查，仍然执行 `ObjectWriterProvider.isNotReferenceDetect` 的 class 排除并忽略 `null`；且压栈/弹栈两侧的排除条件完全一致——`setPath0` 没有压入的值，`popPath0` 也不会弹出。

</details>
